cache_utils/group_backend.py

```diff
@@ -41,8 +41,8 @@ def add(self, key, value, timeout=0, group=None):
 
     def get(self, key, version=None, default=None, group=None):
         key = self._make_key(group, key)
-        packed_value = super(CacheClass, self).get(key, default)
-        if packed_value is None:
+        packed_value = super(CacheClass, self).get(key, default=MISSING)
+        if packed_value is MISSING:
             return default
```

Comment:
so through this, we want to be able to pass `None`, but do we want that stored in the backend? I assume yes, just like any python primitive value that can be pickled. So second time you try to call function you expect the cache to HIT and to return you `None` just like it would return you an integer. Right?

But I looked into how memcache and rediscache both handle this, and they all treat `None` as the reference value that tells them that the key doesn't have a value stored in cache... And `has_key` is implemented the same way, by comparing if `get` returned `None` or not.. See

https://github.com/django/django/blob/950171d7b20a148a964281aed16aff3d2c734ab4/django/core/cache/backends/memcached.py#L72-L73

and

https://github.com/niwinz/django-redis/blob/5d4c737bc5004d7c9ac2bb9951a0e42c7486264d/django_redis/client/default.py#L206-L207

So I think through this change we don't change much unless we also change all the backends we use with cached. Seems overkill for what this achieves if I understood the problem correctly. I'd rather we changed our code that's wrapped by `cached` to return sth other than `None`. Maybe empty string? Thoughts?

Reply:
Great find! Interesting. I believe for the `django-redis` cache, the `None` check is applied to a serialized payload; i.e. it checks if a value is `None` or a bytestring that contains serialized data. It should be possible for that serialized data to deserialize into None after the check.

However, looks like the memcache backend does not allow storing `None`. I suppose it makes sense for Django given that none of its own cache use cases need to distinguish between a missing cache value and a `None`.

For the `cached` decorator that caches the result of any function, I'd consider it a better to design to cache all possible values returned. Otherwise functions must know about this None case and handle it. You can't take any function anymore and slap on a `cached` decorator and trust it to do the right thing, you need to check for any `None` values being returned. I suppose there are counter-arguments to this, but this kind of code is pretty common:

```python
@cached(3600)
def get_most_recent_obj(status):
    return SomeModel.objects.filter(status=status).order_by('-created').first()
```

Reading that declaratively, I wouldn't expect the DB query to be rerun each time if that object doesn't exist (and returns `None`). In this context, `None` means that the expensive computation resulted in no match, rather than that we have no value to cache.

To avoid caching  a value, the user can raise an exception like `DontCacheException` or something more semantically fitting and catch it in their own code, which would prevent the value from being cached.

Since we're using `pickle` to serialize values, we can work around this by having `cached` save a separate sentinel value that represents `None` at the cache level, and replace it with the real `None` for the public interface.

Reply:
> To avoid caching a value, the user can raise an exception like `DontCacheException` or something more semantically fitting and catch it in their own code, which would prevent the value from being cached.

Yep, I was also thinking about sth similar, since sometimes you might not want to cache the thing and just returning `None` is not explicit enough.

> I believe for the django-redis cache, the None check is applied to a serialized payload

That's true!

> Since we're using pickle to serialize values, we can work around this by having cached save a separate sentinel value that represents None at the cache level, and replace it with the real None for the public interface.

Yep, great idea! This should work, as long as the pickling/unpickling behaves as a singleton with several instances of that class. If that's true, then different processes/threads etc. should recognize that singleton value and return `None` to the public interface 👍 .